src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,153 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private static final Object[] EMPTY_ELEMENTDATA = {};
+    private static final int NEGATIVE_INDEX_MARKER = -1;
+    private Object[] elementData;
+    private int size;
+
+    public ArrayList(int initialCapacity) {
+        if (initialCapacity > 0) {
+            this.elementData = new Object[initialCapacity];
+        } else if (initialCapacity == 0) {
+            this.elementData = EMPTY_ELEMENTDATA;
+        } else {
+            throw new IllegalArgumentException("Illegal Capacity: "
+                    + initialCapacity);
+        }
+    }
+
+    public ArrayList() {
+        this.elementData = DEFAULTCAPACITY_EMPTY_ELEMENTDATA;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size < this.elementData.length) {
+            elementData[size] = value;
+            size++;
+        } else {
+            add(value, size);
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        rangeCheckForAdd(index);
+        if (size == this.elementData.length) {
+            elementData = growIfArrayFull();
+        }
+        System.arraycopy(elementData, index,
+                elementData, index + 1,
+                    size - index);
+        elementData[index] = value;
+        size++;
 
     }
 
+    private void rangeCheckForAdd(int index) {
```

Comment:
Don't create repeating code. The index check logic is repeated in multiple methods. Consider refactoring it into a separate method to avoid duplication.